core/validation.go

```diff
@@ -166,3 +180,139 @@ func (app *App) orderAlreadyStored(orderHash common.Hash) (bool, error) {
 	}
 	return false, err
 }
+
+// TODO(albrow): This function needs to be rigorously tested.
+func (app *App) validateETHBacking(orders []*zeroex.SignedOrder) (ordersWithSufficientBacking []*zeroex.SignedOrder, rejectedOrders []*zeroex.RejectedOrderInfo) {
+	totalExistingOrders, err := app.db.Orders.Count()
+	if err != nil {
+		return nil, nil
+	}
+	if totalExistingOrders+len(orders) <= app.config.MaxOrdersInStorage {
+		// If we haven't yet reached our storage limit, all orders are considered
+		// valid in terms of their ETH backing.
+		return orders, nil
+	}
+
+	// Group orders by their maker address.
+	ordersByMakerAddress := map[common.Address][]*zeroex.SignedOrder{}
+	for _, order := range orders {
+		if _, found := ordersByMakerAddress[order.MakerAddress]; found {
+			ordersByMakerAddress[order.MakerAddress] = append(ordersByMakerAddress[order.MakerAddress], order)
+		} else {
+			ordersByMakerAddress[order.MakerAddress] = []*zeroex.SignedOrder{order}
+		}
+	}
+	var makerAddresses []common.Address
+	for makerAddress := range ordersByMakerAddress {
+		makerAddresses = append(makerAddresses, makerAddress)
+	}
+
+	// Get the current ETH balance for each maker.
+	addressToBalance, failedAddresses := app.ethWathcher.Add(makerAddresses)
```

Comment:
I think we are going to end up with the conclusion that anything that is rebroadcast will need to be stored in some way. Otherwise we will be able to suck up bandwidth by repeatedly forcing rebroadcast of the same set of bogus messages. It doesn't need to be the full message, just a hash will do.

Reply:
> This is a bit strange. We are going to be tracking makerAddresses, even of orders that won't ultimately get stored... Is that desirable?

@fabioberger I think it is desirable. Even in the current implementation which doesn't store anything for invalid orders, caching these balances will make it more efficient to check incoming orders from makers we have already seen.

> Also, `ethWathcher.Add` as it's currently written only returns the balances of previously unknown ETH addresses passed in. We should probably refactor so it returns balances for all passed in addresses.

@fabioberger good point. Yeah we will need to address this.

> I think we are going to end up with the conclusion that anything that is rebroadcast will need to be stored in some way. Otherwise we will be able to suck up bandwidth by repeatedly forcing rebroadcast of the same set of bogus messages. It doesn't need to be the full message, just a hash will do.

@0xpeter we *probably* want to do this to some extent but not in this PR.